File: web/src/utils/k8sTolerations.ts
export const MAX_K8S_DS_TOLERATIONS = 16;
export const K8S_DS_TOLERATION_EFFECTS = ['NoSchedule', 'NoExecute'] as const;

export const DEFAULT_K8S_DS_TOLERATIONS: K8sToleration[] = [
  { key: 'node-role.kubernetes.io/control-plane', effect: 'NoSchedule' },
  { key: 'node-role.kubernetes.io/master', effect: 'NoSchedule' }
];

export type K8sTolerationEffect = (typeof K8S_DS_TOLERATION_EFFECTS)[number];

export interface K8sToleration {
  key: string;
  effect: K8sTolerationEffect;
  value?: string;
}

export type K8sTolerationMode = 'default' | 'custom' | 'none';

const NAME_RE = /^[A-Za-z0-9]([A-Za-z0-9._-]{0,61}[A-Za-z0-9])?$/;
const DNS_LABEL_RE = /^[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?$/;
const ALLOWED_FIELDS = new Set(['key', 'effect', 'value']);

const isValidKey = (key: unknown): key is string => {
  if (typeof key !== 'string' || !key || key.includes('__') || (key.match(/\//g) || []).length > 1) {
    return false;
  }
  if (key.includes('/')) {
    const [prefix, name] = key.split('/');
    if (
      prefix.length > 253 ||
      !prefix.split('.').every((part) => DNS_LABEL_RE.test(part)) ||
      !NAME_RE.test(name)
    ) {
      return false;
    }
    return true;
  }
  return NAME_RE.test(key);
};

const isValidValue = (value: unknown): value is string => {
  if (typeof value !== 'string' || value.includes('__')) {
    return false;
  }
  return !value || NAME_RE.test(value);
};

const isValidItem = (entry: unknown): entry is K8sToleration => {
  if (!entry || typeof entry !== 'object' || Array.isArray(entry)) {
    return false;
  }
  const item = entry as Record<string, unknown>;
  if (Object.keys(item).some((field) => !ALLOWED_FIELDS.has(field))) {
    return false;
  }
  if (!isValidKey(item.key)) {
    return false;
  }
  if (item.effect !== 'NoSchedule' && item.effect !== 'NoExecute') {
    return false;
  }
  if (item.value !== undefined && item.value !== null && !isValidValue(item.value)) {
    return false;
  }
  return true;
};

export const isValidK8sTolerations = (value: unknown): boolean => {
  if (value == null) {
    return true;
  }
  if (!Array.isArray(value) || value.length > MAX_K8S_DS_TOLERATIONS) {
    return false;
  }
  return value.every(isValidItem);
};

export const normalizeK8sTolerations = (
  value: unknown
): K8sToleration[] | null => {
  if (value == null) {
    return null;
  }
  if (!isValidK8sTolerations(value) || !Array.isArray(value)) {
    throw new Error('invalid k8s tolerations');
  }
  return (value as K8sToleration[]).map((entry) => {
    const item: K8sToleration = {
      key: entry.key,
      effect: entry.effect
    };
    if (typeof entry.value === 'string' && entry.value !== '') {
      item.value = entry.value;
    }
    return item;
  });
};

export const k8sTolerationModeFromValue = (
  value: unknown
): K8sTolerationMode => {
  if (Array.isArray(value) && value.length === 0) {
    return 'none';
  }
  if (Array.isArray(value)) {
    return 'custom';
  }
  return 'default';
};

export const k8sTolerationsFromMode = (
  mode: K8sTolerationMode,
  items?: unknown
): K8sToleration[] | null => {
  if (mode === 'none') {
    return [];
  }
  if (mode === 'custom') {
    return normalizeK8sTolerations(items ?? []);
  }
  return null;
};

File: server/apps/core/utils/k8s_tolerations.py
import re

from apps.core.exceptions.base_app_exception import ValidationAppException

MAX_K8S_DS_TOLERATIONS = 16
ALLOWED_K8S_DS_TOLERATION_EFFECTS = frozenset({"NoSchedule", "NoExecute"})
ALLOWED_K8S_DS_TOLERATION_FIELDS = frozenset({"key", "effect", "value"})
DEFAULT_K8S_DS_TOLERATIONS = [
    {"key": "node-role.kubernetes.io/control-plane", "effect": "NoSchedule"},
    {"key": "node-role.kubernetes.io/master", "effect": "NoSchedule"},
]

_NAME_RE = re.compile(r"^[A-Za-z0-9]([A-Za-z0-9._-]{0,61}[A-Za-z0-9])?$")
_DNS_LABEL_RE = re.compile(r"^[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?$")


def _validate_key(key) -> str:
    if not isinstance(key, str) or not key:
        raise ValidationAppException("污点容忍 key 必填且必须为非空字符串")
    if "__" in key:
        raise ValidationAppException("污点容忍 key 不能包含 __")
    if key.count("/") > 1:
        raise ValidationAppException("污点容忍 key 最多包含一个 /")
    if "/" in key:
        prefix, name = key.split("/")
        if len(prefix) > 253 or not all(_DNS_LABEL_RE.fullmatch(part) for part in prefix.split(".")):
            raise ValidationAppException("污点容忍 key 前缀不是合法 DNS 子域")
    else:
        name = key
    if not _NAME_RE.fullmatch(name):
        raise ValidationAppException("污点容忍 key 不符合 Kubernetes qualified name")
    return key


def _validate_value(value) -> str:
    if not isinstance(value, str):
        raise ValidationAppException("污点容忍 value 必须为字符串")
    if "__" in value:
        raise ValidationAppException("污点容忍 value 不能包含 __")
    if value and not _NAME_RE.fullmatch(value):
        raise ValidationAppException("污点容忍 value 不符合 Kubernetes label value")
    return value


def _normalize_item(entry) -> dict:
    if not isinstance(entry, dict):
        raise ValidationAppException("污点容忍清单项必须为对象")
    unknown = sorted(set(entry) - ALLOWED_K8S_DS_TOLERATION_FIELDS)
    if unknown:
        raise ValidationAppException("污点容忍清单不允许未知字段")

    item = {
        "key": _validate_key(entry.get("key")),
        "effect": entry.get("effect"),
    }
    if item["effect"] not in ALLOWED_K8S_DS_TOLERATION_EFFECTS:
        raise ValidationAppException("污点容忍 effect 必须为 NoSchedule 或 NoExecute")

    if "value" in entry and entry.get("value") is not None:
        item["value"] = _validate_value(entry.get("value"))
    return item


def normalize_k8s_tolerations(value=None):
    """规范化 DaemonSet 污点容忍清单。

    None 表示未配置，调用方不得传给 webhookd（缺省注入默认两条）。
    空列表表示显式零容忍，必须原样下发。
    """
    if value is None:
        return None
    if not isinstance(value, list):
        raise ValidationAppException("污点容忍清单必须为数组")
    if len(value) > MAX_K8S_DS_TOLERATIONS:
        raise ValidationAppException(f"污点容忍清单最多 {MAX_K8S_DS_TOLERATIONS} 项")
    return [_normalize_item(entry) for entry in value]


def apply_k8s_tolerations_param(params: dict, value=None) -> dict:
    """把三态清单写入 webhookd 请求：None 省略字段，[] 显式下发。"""
    normalized = normalize_k8s_tolerations(value)
    if normalized is not None:
        params["tolerations"] = normalized
    return params
